# Regression Week 2: Multiple Regression (Interpretation)

The goal of this first notebook is to explore multiple regression and feature engineering.

In this notebook you will use data on house sales in King County to predict prices using multiple regression. You will:
* Use **Pandas/Numpy** to do some feature engineering
* Use **Scikit-learn** to compute the regression weights (coefficients/parameters)
* Given the regression weights, predictors and outcome write a function to compute the Residual Sum of Squares
* Look at coefficients and interpret their meanings
* Evaluate multiple models via RSS

# Imports

In [1]:
# Standard libs
import pandas as pd
import numpy as np

In [3]:
# Machine Learning
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [4]:
from math import log

# Load in house sales data

Dataset is from house sales in King County, the region where the city of Seattle, WA is located.

In [5]:
house_sales_datatype={'bathrooms':float, 'waterfront':int, 'sqft_above':int,
                      'sqft_living15':float, 'grade':int, 'yr_renovated':int,
                      'price':float, 'bedrooms':float, 'zipcode':str, 'long':float,
                      'sqft_lot15':float, 'sqft_living':float, 'floors':str, 'condition':int,
                      'lat':float, 'date':str, 'sqft_basement':int, 'yr_built':int, 'id':str,
                      'sqft_lot':int, 'view':int}

In [6]:
sales = pd.read_csv("./data/kc_house_data.csv", dtype=house_sales_datatype)

In [7]:
sales.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0       3.0       1.00       1180.0   
1  6414100192  20141209T000000   538000.0       3.0       2.25       2570.0   
2  5631500400  20150225T000000   180000.0       2.0       1.00        770.0   
3  2487200875  20141209T000000   604000.0       4.0       3.00       1960.0   
4  1954400510  20150218T000000   510000.0       3.0       2.00       1680.0   
5  7237550310  20140512T000000  1225000.0       4.0       4.50       5420.0   
6  1321400060  20140627T000000   257500.0       3.0       2.25       1715.0   
7  2008000270  20150115T000000   291850.0       3.0       1.50       1060.0   
8  2414600126  20150415T000000   229500.0       3.0       1.00       1780.0   
9  3793500160  20150312T000000   323000.0       3.0       2.50       1890.0   

   sqft_lot floors  waterfront  view     ...      grade  sqft_above  \
0      5650      1           0     0     ...          7        1180   
1      7242      2           0     0     ...          7        2170   
2     10000      1           0     0     ...          6         770   
3      5000      1           0     0     ...          7        1050   
4      8080      1           0     0     ...          8        1680   
5    101930      1           0     0     ...         11        3890   
6      6819      2           0     0     ...          7        1715   
7      9711      1           0     0     ...          7        1060   
8      7470      1           0     0     ...          7        1050   
9      6560      2           0     0     ...          7        1890   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   
5           1530      2001             0    98053  47.6561 -122.005   
6              0      1995             0    98003  47.3097 -122.327   
7              0      1963             0    98198  47.4095 -122.315   
8            730      1960             0    98146  47.5123 -122.337   
9              0      2003             0    98038  47.3684 -122.031   

   sqft_living15  sqft_lot15  
0         1340.0      5650.0  
1         1690.0      7639.0  
2         2720.0      8062.0  
3         1360.0      5000.0  
4         1800.0      7503.0  
5         4760.0    101930.0  
6         2238.0      6819.0  
7         1650.0      9711.0  
8         1780.0      8113.0  
9         2390.0      7570.0  

[10 rows x 21 columns]

# Load training and test data
Load training and test data from the files provided with the course assignment.  

In [8]:
train_data = pd.read_csv("./data/kc_house_train_data.csv", dtype=house_sales_datatype)
test_data = pd.read_csv("./data/kc_house_test_data.csv", dtype=house_sales_datatype)

# Learning a multiple regression model

Using Sciki-learn, fit a multiple regression model on training data predicting 'price' based on the following features:

example_features = **['sqft_living', 'bedrooms', 'bathrooms']**

In [9]:
example_features = ["sqft_living", "bedrooms", "bathrooms"]
X_train = train_data[example_features]
y_train = train_data["price"]

In [10]:
example_model = LinearRegression()
example_model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

Now that we have fitted the model we can extract the regression weights (coefficients)

In [11]:
example_weight_summary = np.append(example_model.intercept_,example_model.coef_)
print("example_weight_summary:",example_weight_summary)

example_weight_summary: [ 87912.86581496    315.40669062 -65081.88711588   6942.16598637]


# Making Predictions

Once a model is built we can use the .predict() function to find the predicted values for data we pass. For example using the example model above:

In [12]:
example_predictions = example_model.predict(X_train)
print("example_predictions[0]", example_predictions[0]) # should be 271789.505878

example_predictions[0] 271789.26538


# Compute RSS

Now that we can make predictions given the model, let's write a function to compute the RSS of the model. Complete the function below to calculate RSS given the model, data, and the outcome.

In [13]:
def get_residual_sum_of_squares(model, data, outcome):
    # First get the predictions
    y_predict = model.predict(data) 

    # Compute RSS
    RSS = (np.square(outcome - y_predict).sum())

    return(RSS)    

Test your function by computing the RSS on **TEST** data for the example model:

In [14]:
X_test = test_data[example_features]
y_test = test_data["price"]

In [15]:
rss_example_train = get_residual_sum_of_squares(example_model, X_test, y_test)
print("rss_example_train:", rss_example_train) # should be 2.7376153833e+14

rss_example_train: 273761940583133.97


**Comparing our function with scikit-learn mean_squared_error**
scikit-learn mean_square_error = RSS/N

with N=number of observations

In [27]:
mse = mean_squared_error(y_test,example_model.predict(X_test),multioutput="uniform_average")
msebyN = mse*y_test.shape[0]
print("mean_squared_error (scikit-learn):",round(mse,2))
print("mean_squared_error * N number of observations:",round(msebyN,2))

mean_squared_error (scikit-learn): 64734438539.4
mean_squared_error * N number of observations: 2.73761940583e+14


# Create some new features

Although we often think of multiple regression as including multiple different features (e.g. **# of bedrooms, squarefeet, and # of bathrooms**), we can also consider transformations of existing features e.g. the log of the squarefeet or even "interaction" features such as the product of bedrooms and bathrooms.

Next create the following 4 new features as column in both TEST and TRAIN data:
* bedrooms_squared = bedrooms\*bedrooms
* bed_bath_rooms = bedrooms\*bathrooms
* log_sqft_living = log(sqft_living)
* lat_plus_long = lat + long 
As an example here's the first one:

In [16]:
# bedrooms_squared
train_data["bedrooms_squared"] = train_data["bedrooms"].apply(lambda x: x**2)
test_data["bedrooms_squared"] = test_data["bedrooms"].apply(lambda x: x**2)

# bed_bath_rooms
train_data["bed_bath_rooms"] = train_data["bedrooms"] * train_data["bathrooms"]
test_data["bed_bath_rooms"] = test_data["bedrooms"] * test_data["bathrooms"]

# log_sqft_living
train_data["log_sqft_living"] = train_data["sqft_living"].apply(lambda x: log(x))
test_data["log_sqft_living"] = test_data["sqft_living"].apply(lambda x: log(x))

#lat_plus_long
train_data["lat_plus_long"] = train_data["lat"] + train_data["long"]
test_data["lat_plus_long"] = test_data["lat"] + test_data["long"]

In [17]:
train_data.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0       3.0       1.00       1180.0   
1  6414100192  20141209T000000   538000.0       3.0       2.25       2570.0   
2  5631500400  20150225T000000   180000.0       2.0       1.00        770.0   
3  2487200875  20141209T000000   604000.0       4.0       3.00       1960.0   
4  1954400510  20150218T000000   510000.0       3.0       2.00       1680.0   
5  7237550310  20140512T000000  1225000.0       4.0       4.50       5420.0   
6  1321400060  20140627T000000   257500.0       3.0       2.25       1715.0   
7  2008000270  20150115T000000   291850.0       3.0       1.50       1060.0   
8  2414600126  20150415T000000   229500.0       3.0       1.00       1780.0   
9  3793500160  20150312T000000   323000.0       3.0       2.50       1890.0   

   sqft_lot floors  waterfront  view      ...        yr_renovated  zipcode  \
0      5650      1           0     0      ...                   0    98178   
1      7242      2           0     0      ...                1991    98125   
2     10000      1           0     0      ...                   0    98028   
3      5000      1           0     0      ...                   0    98136   
4      8080      1           0     0      ...                   0    98074   
5    101930      1           0     0      ...                   0    98053   
6      6819      2           0     0      ...                   0    98003   
7      9711      1           0     0      ...                   0    98198   
8      7470      1           0     0      ...                   0    98146   
9      6560      2           0     0      ...                   0    98038   

       lat     long  sqft_living15  sqft_lot15 bedrooms_squared  \
0  47.5112 -122.257         1340.0      5650.0              9.0   
1  47.7210 -122.319         1690.0      7639.0              9.0   
2  47.7379 -122.233         2720.0      8062.0              4.0   
3  47.5208 -122.393         1360.0      5000.0             16.0   
4  47.6168 -122.045         1800.0      7503.0              9.0   
5  47.6561 -122.005         4760.0    101930.0             16.0   
6  47.3097 -122.327         2238.0      6819.0              9.0   
7  47.4095 -122.315         1650.0      9711.0              9.0   
8  47.5123 -122.337         1780.0      8113.0              9.0   
9  47.3684 -122.031         2390.0      7570.0              9.0   

   bed_bath_rooms  log_sqft_living  lat_plus_long  
0            3.00         7.073270       -74.7458  
1            6.75         7.851661       -74.5980  
2            2.00         6.646391       -74.4951  
3           12.00         7.580700       -74.8722  
4            6.00         7.426549       -74.4282  
5           18.00         8.597851       -74.3489  
6            6.75         7.447168       -75.0173  
7            4.50         6.966024       -74.9055  
8            3.00         7.484369       -74.8247  
9            7.50         7.544332       -74.6626  

[10 rows x 25 columns]

* Squaring bedrooms will increase the separation between not many bedrooms (e.g. 1) and lots of bedrooms (e.g. 4) since 1^2 = 1 but 4^2 = 16. Consequently this feature will mostly affect houses with many bedrooms.
* bedrooms times bathrooms gives what's called an "interaction" feature. It is large when *both* of them are large.
* Taking the log of squarefeet has the effect of bringing large values closer together and spreading out small values.
* Adding latitude to longitude is totally non-sensical but we will do it anyway (you'll see why)

**Quiz Question: What is the mean (arithmetic average) value of your 4 new features on TEST data? (round to 2 digits)**

In [19]:
print("Mean bedrooms_squared on TEST data:",round(test_data["bedrooms_squared"].mean(),2))
print("Mean bed_bath_rooms on TEST data:",round(test_data["bed_bath_rooms"].mean(),2))
print("Mean log_sqft_living on TEST data:",round(test_data["log_sqft_living"].mean(),2))
print("Mean lat_plus_long on TEST data:",round(test_data["lat_plus_long"].mean(),2))

Mean bedrooms_squared on TEST data: 12.45
Mean bed_bath_rooms on TEST data: 7.5
Mean log_sqft_living on TEST data: 7.55
Mean lat_plus_long on TEST data: -74.65


# Learning Multiple Models

Now we will learn the weights for three (nested) models for predicting house prices. The first model will have the fewest features the second model will add one more feature and the third will add a few more:
* Model 1: squarefeet, # bedrooms, # bathrooms, latitude & longitude
* Model 2: add bedrooms\*bathrooms
* Model 3: Add log squarefeet, bedrooms squared, and the (nonsensical) latitude + longitude

In [20]:
model_1_features = ["sqft_living", "bedrooms", "bathrooms", "lat", "long"]
model_2_features = model_1_features + ["bed_bath_rooms"]
model_3_features = model_2_features + ["bedrooms_squared", "log_sqft_living", "lat_plus_long"]

Now that you have the features, learn the weights for the three different models and look at the value of the weights/coefficients.

### Model 1

In [21]:
X_train_model_1 = train_data[model_1_features]
y_train = train_data["price"]

model_1 = LinearRegression()
model_1.fit(X_train_model_1, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

### Model 2

In [22]:
X_train_model_2 = train_data[model_2_features]

model_2 = LinearRegression()
model_2.fit(X_train_model_2, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

### Model 3

In [23]:
X_train_model_3 = train_data[model_3_features]

model_3 = LinearRegression()
model_3.fit(X_train_model_3, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

Get coefficients for all models

In [24]:
print("coefficients for model_1:",np.append(model_1.intercept_,model_1.coef_))
print("coefficients for model_2:",np.append(model_2.intercept_,model_2.coef_))
print("coefficients for model_3:",np.append(model_3.intercept_,model_3.coef_))

coefficients for model_1: [ -6.90757268e+07   3.12258646e+02  -5.95865332e+04   1.57067421e+04
   6.58619264e+05  -3.09374351e+05]
coefficients for model_2: [ -6.68679689e+07   3.06610053e+02  -1.13446368e+05  -7.14613083e+04
   6.54844630e+05  -2.94298969e+05   2.55796520e+04]
coefficients for model_3: [ -6.20360850e+07   5.29422820e+02   3.45142296e+04   6.70607813e+04
   5.34085611e+05  -4.06750711e+05  -8.57050439e+03  -6.78858667e+03
  -5.61831484e+05   1.27334900e+05]


**Quiz Question: What is the sign (positive or negative) for the coefficient/weight for 'bathrooms' in model 1?**

bathrooms is **third** features so associated with **fourth** coefficient, here for **model_1: 1.57** so **POSITIVE**

**Quiz Question: What is the sign (positive or negative) for the coefficient/weight for 'bathrooms' in model 2?**

bathrooms is **third** features so associated with **fourth** coefficient, here for **model_2: -7.14** so **NEGATIVE**

# Comparing multiple models

Now that you've learned three models and extracted the model weights we want to evaluate which model is best.

First use your functions from earlier to compute the RSS on TRAINING Data for each of the three models.

In [62]:
RSS = pd.DataFrame([get_residual_sum_of_squares(model_1, X_train_model_1, y_train),
                    get_residual_sum_of_squares(model_2, X_train_model_2, y_train),
                    get_residual_sum_of_squares(model_3, X_train_model_3, y_train)],
                   index=["model_1","model_2","model_3"],
                   columns=["RSS_train"],
                   dtype=float)

In [63]:
RSS

RSS_train
model_1  9.678800e+14
model_2  9.584196e+14
model_3  9.034365e+14

**Quiz Question: Which model (1, 2 or 3) has lowest RSS on TRAINING Data?** Is this what you expected?

In [64]:
RSS["RSS_train"].idxmin()

'model_3'

Now compute the RSS on on TEST data for each of the three models.

In [65]:
X_test_model_1 = test_data[model_1_features]
X_test_model_2 = test_data[model_2_features]
X_test_model_3 = test_data[model_3_features]

y_test = test_data["price"]

In [66]:
RSS["RSS_test"] = [get_residual_sum_of_squares(model_1, X_test_model_1, y_test),
                    get_residual_sum_of_squares(model_2, X_test_model_2, y_test),
                    get_residual_sum_of_squares(model_3, X_test_model_3, y_test)]

In [67]:
RSS

RSS_train      RSS_test
model_1  9.678800e+14  2.255005e+14
model_2  9.584196e+14  2.233775e+14
model_3  9.034365e+14  2.592363e+14

**Quiz Question: Which model (1, 2 or 3) has lowest RSS on TESTING Data?** Is this what you expected?Think about the features that were added to each model from the previous.

In [68]:
RSS["RSS_test"].idxmin()

'model_2'